internal/wrphandlers/qos/qos.go

```diff
@@ -0,0 +1,161 @@
+// SPDX-FileCopyrightText: 2024 Comcast Cable Communications Management, LLC
+// SPDX-License-Identifier: Apache-2.0
+
+package qos
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/xmidt-org/wrp-go/v3"
+	"github.com/xmidt-org/xmidt-agent/internal/websocket"
+	"github.com/xmidt-org/xmidt-agent/internal/wrpkit"
+)
+
+var (
+	ErrInvalidInput     = errors.New("invalid input")
+	ErrMisconfiguredQOS = errors.New("misconfigured QOS")
+)
+
+// Option is a functional option type for QOS.
+type Option interface {
+	apply(*Handler) error
+}
+
+type optionFunc func(*Handler) error
+
+func (f optionFunc) apply(c *Handler) error {
+	return f(c)
+}
+
+// Handler queues incoming messages and sends them to the next wrphandler
+type Handler struct {
+	next wrpkit.Handler
+	// queue for wrp messages
+	queue PriorityQueue
+
+	m sync.Mutex
+	// ingestingQueue blocks if the SendQueuedMessages() is already running, i.e.: send
```

Comment:
> None of these (ingestingQueue, shutdown, ctx) fields are needed

in terms of `ingestingQueue`, I was trying to limit how many parallel `go h.SendQueuedMessages()`  goroutines can be ran, since our websocket locks during writes. From my understanding that meant we can potentially have the following problems:
- unbounded growth of `go h.SendQueuedMessages()`  goroutines
- a group of goroutines waiting to write to the ws one at time, defeating the purpose of having more than 1 goroutine
 

>Since the queue is added to or removed from, then unlocked in between, it gives the cleanup process the ability to grab the mutex & dump everything right then. 

what triggers the cleanup process? I'm assuming this is related to this comment as well https://github.com/xmidt-org/xmidt-agent/pull/95/files#r1571847953:

> 
> Stop() should change to something like Cleanup() or Empty() or Drain() because that's all it's going to do ... just throw away everything in the PQ.
> 
> Instead of calling shutdown(), just rapid fire Pop() everything off the heap, or replace the heap with an empty one. Do this in the mutex lock region.
> Comment
> 

I thought the cleanup process, dump everything without attempting to send the messages, would naturally be the go program being restarted. 

right now `Stop` is called during a disconnect event so that we stop the running `go h.SendQueuedMessages()` goroutine with `shutdown, ctx`, but we keep the queue as is and wait until the next reconnect event to start ingesting the queue again (starting where we left off). 

I assumed that's what we wanted (to hold onto messages between disconnects) since that's the only reason (from my understanding) why messages would fail to send

That's why I used `shutdown, ctx` as the stopping mechanism for the queue rather them dumping everything